pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java

```diff
@@ -111,12 +114,34 @@ public static boolean containsFunction(String functionName) {
    * methods are already registered.
    */
   @Nullable
-  public static FunctionInfo getFunctionInfo(String functionName, int numParameters) {
-    Map<Integer, FunctionInfo> functionInfoMap = FUNCTION_INFO_MAP.get(canonicalize(functionName));
-    return functionInfoMap != null ? functionInfoMap.get(numParameters) : null;
+  public static FunctionInfo getFunctionInfo(String functionName, String[] paramClassName) {
+    String paramTypeString = getSortedParamClassNameString(paramClassName);
+    Map<String, FunctionInfo> functionInfoMap = FUNCTION_INFO_MAP.get(canonicalize(functionName));
+    return functionInfoMap != null ? functionInfoMap.get(paramTypeString) : null;
```

Comment:
If there is no match, we should return one with the same number of parameters.
Currently the arguments will be automatically casted if the class does not match. We need to keep it backward-compatible.

Reply:
+1

Also, when multiple ones available with the same number of parameters, which one to pick?

Reply:
This would also need a way to define functions for different types without repeating the same logic again and again, right? For example `ArithmeticFunctions.java` currently defines:
```
  public static double plus(double a, double b);
```
Defining this function for all combination of numerical types (int, long, float, double) means writing the `plus` function 16 times! One way is to use some sort of compile time templating solution (something like in C++). An alternating solution might be to create a `FunctionArg` object and then write functions to take `FunctionArg` as input parameter and produce a `FunctionArg` as output parameter. This way a function can be written only once and be able to handle different types.

While this PR is definitely a step in the right direction, not sure how much of a value this PR will provide without surrounding logic and support. I think this is something for which we need a design doc so that a couple of alternatives can be considered and the issue can be considered holistically before narrowing down on the solution.